File: api-php/samples/server/src/advanced.php
<?php

use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;
use Trinsic\Api\Model\CreateAdvancedProviderSessionRequest as CreateAdvancedProviderSessionRequest;
use Trinsic\Api\Model\CreateAdvancedProviderSessionResponse as CreateAdvancedProviderSessionResponse;
use Trinsic\Api\Model\GetSessionResultRequest as GetSessionResultRequest;
use Trinsic\Api\Model\RefreshStepContentRequest as RefreshStepContentRequest;
use Trinsic\Api\ApiException as ApiException;

return function ($app, $sessions) {
    
    // Index route to serve the index.html file
    $app->get('/advanced', function (Request $request, Response $response, $args) {
        return $response
            ->withHeader('Location', '/advanced.html')
            ->withStatus(302);
    });

    $app->get('/advanced-popup', function (Request $request, Response $response, $args) {
        $queryParams = $request->getQueryParams();
        
        $queryString = http_build_query($queryParams);
        
        $redirectUrl = '/advanced-popup.html' . (!empty($queryString) ? '?' . $queryString : '');
        return $response
            ->withHeader('Location', $redirectUrl)
            ->withStatus(302);
    });

    $app->get("/advanced-launch/{provider}", function (Request $request, Response $response, $args) use ($sessions) {
        try {
            $provider = $args['provider'];
            $queryParams = $request->getQueryParams();
            $redirectUrl = $queryParams['redirectUrl'] ?? null;
            $fallbackToTrinsicUI = ($queryParams['fallbackToTrinsicUI'] ?? 'false') === 'true';
            $capabilities = isset($queryParams['capabilities']) ? explode(",", $queryParams['capabilities']) : [];
    
            $req = new CreateAdvancedProviderSessionRequest();
            $req->setRedirectUrl($redirectUrl);
            $req->setProvider($provider);
            $req->setCapabilities($capabilities);
            $req->setFallbackToHostedUi($fallbackToTrinsicUI);
    
            $result = $sessions->createAdvancedProviderSession($req);
    
            if ($result->getNextStep()->getMethod() === 'LaunchBrowser') {
                return $response
                    ->withHeader('Location', $result->getNextStep()->getContent())
                    ->withStatus(302);
            } else {
                $shouldRefresh = !is_null($result->getNextStep()->getRefresh());
                $refreshAfter = $shouldRefresh ? $result->getNextStep()->getRefresh()->getRefreshAfter()->format(DATE_ISO8601) : gmdate(DATE_ISO8601);
    
                $queryParams = http_build_query([
                    'sessionId' => $result->getSessionId(),
                    'resultsAccessKey' => $result->getResultCollection()->getResultsAccessKey(),
                    'nextStep' => $result->getNextStep()->getMethod(),
                    'content' => $result->getNextStep()->getContent(),
                    'shouldRefresh' => $shouldRefresh ? 'true' : 'false',
                    'refreshAfter' => $refreshAfter
                ]);
    
                return $response
                    ->withHeader('Location', "/advanced-popup?$queryParams")
                    ->withStatus(302);
            }
        } catch (ApiException $e) {
            $errorContent = urlencode($e->getResponseBody());
            return $response
                ->withHeader('Location', "/advanced-popup?error=$errorContent")
                ->withStatus(302);
        }
    });
    
    $app->post("/refresh-content/{sessionId}", function (Request $request, Response $response, $args) use ($sessions) {
        $sessionId = $args['sessionId'];
        $body = json_decode($request->getBody()->getContents(), true);
        $resultsAccessKey = $body['resultsAccessKey'] ?? null;
    
        $req = new RefreshStepContentRequest();
        $req->setResultsAccessKey($resultsAccessKey);
        $result = $sessions->refreshStepContent($sessionId, $req);
    
        $response->getBody()->write(json_encode($result));
        return $response->withHeader('Content-Type', 'application/json');
    });
    
    $app->post("/poll-results/{sessionId}", function (Request $request, Response $response, $args) use ($sessions) {
        $sessionId = $args['sessionId'];
        $body = json_decode($request->getBody()->getContents(), true);
        $resultsAccessKey = $body['resultsAccessKey'] ?? null;
    
        $req = new GetSessionResultRequest();
        $req->setResultsAccessKey($resultsAccessKey);
    
        $result = $sessions->getSessionResult($sessionId, $req);
    
        $response->getBody()->write(json_encode($result));
        return $response->withHeader('Content-Type', 'application/json');
    });

};

File: api-java/samples/server/trinsic/Advanced.java
package id.trinsic;

import id.trinsic.api.SessionsApi;
import id.trinsic.api.models.*;
import io.javalin.Javalin;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;

public class Advanced {
    public static void AdvancedRoutes(Javalin app, SessionsApi session){
        app.get("/advanced", ctx -> {
            ctx.redirect("/advanced.html");
        });

        app.get("/advanced-popup", ctx -> {
            String query = ctx.queryString(); // grabs everything after '?'
            String target = "/advanced-popup.html" + (query != null ? "?" + query : "");
            ctx.redirect(target);
        });

        app.get("/advanced-launch/{provider}", ctx -> {
            try {
                String provider = ctx.pathParam("provider");
                String redirectUrl = ctx.queryParam("redirectUrl");
                boolean fallbackToTrinsicUI = "true".equals(ctx.queryParam("fallbackToTrinsicUI"));

                // Build request
                CreateAdvancedProviderSessionRequest req = new CreateAdvancedProviderSessionRequest();
                req.setProvider(provider);
                req.setRedirectUrl(redirectUrl);
                req.setFallbackToHostedUI(fallbackToTrinsicUI);

                String capabilitiesParam = ctx.queryParam("capabilities");
                List<IntegrationCapability> capabilities = capabilitiesParam != null
                        ? Arrays.stream(capabilitiesParam.split(","))
                        .map(String::trim)
                        .map(IntegrationCapability::fromValue)
                        .collect(Collectors.toList())
                        : List.of();

                req.setCapabilities(capabilities);

                // Call API
                CreateAdvancedProviderSessionResponse result =
                        session.createAdvancedProviderSession(req);

                var nextStep = result.getNextStep();
                if (nextStep.getMethod().getValue().equals(IntegrationLaunchMethod.LAUNCH_BROWSER.getValue())) {
                    ctx.redirect(nextStep.getContent());
                } else {
                    boolean shouldRefresh = nextStep.getRefresh() != null;
                    String refreshAfter = shouldRefresh
                            ? nextStep.getRefresh().getRefreshAfter().toString()
                            : Instant.now().toString();

                    String redirectQuery = String.format(
                            "sessionId=%s&resultsAccessKey=%s&nextStep=%s&content=%s&shouldRefresh=%s&refreshAfter=%s",
                            URLEncoder.encode(result.getSessionId().toString(), StandardCharsets.UTF_8),
                            URLEncoder.encode(result.getResultCollection().getResultsAccessKey(), StandardCharsets.UTF_8),
                            URLEncoder.encode(nextStep.getMethod().getValue(), StandardCharsets.UTF_8),
                            URLEncoder.encode(nextStep.getContent(), StandardCharsets.UTF_8),
                            shouldRefresh,
                            URLEncoder.encode(refreshAfter, StandardCharsets.UTF_8)
                    );

                    ctx.redirect("/advanced-popup?" + redirectQuery);
                }
            } catch (ApiException e) {
                String error = URLEncoder.encode(e.getResponseBody(), StandardCharsets.UTF_8);
                ctx.redirect("/advanced-popup?error=" + error);
            }
        });

        app.post("/refresh-content/{sessionId}", ctx -> {
            String sessionId = ctx.pathParam("sessionId");

            Map<String, Object> body = ctx.bodyAsClass(Map.class);
            String resultsAccessKey = (String) body.get("resultsAccessKey");

            RefreshStepContentRequest request = new RefreshStepContentRequest();
            request.setResultsAccessKey(resultsAccessKey);

            RefreshStepContentResponse result =
                    session.refreshStepContent(UUID.fromString(sessionId), request);

            ctx.json(result); // Or wrap it if you need `deep_transform`
        });

        app.post("/poll-results/{sessionId}", ctx -> {
            String sessionId = ctx.pathParam("sessionId");

            Map<String, Object> body = ctx.bodyAsClass(Map.class);
            String resultsAccessKey = (String) body.get("resultsAccessKey");

            GetSessionResultRequest request = new GetSessionResultRequest();
            request.setResultsAccessKey(resultsAccessKey);

            GetSessionResultResponse result =
                    session.getSessionResult(sessionId, request);

            ctx.json(result); // Or wrap if needed
        });
    }
}
